### PR TITLE
Implements a `Iterator` on `Msg` and a `DecodedField` generic enum

### PR DESCRIPTION
**Pull request by pfernie**
This PR adds 2 pieces of functionality; would be happy to split into separate PRs if prefered.

1. Add `Iterator` implemention for `&Msg`, allowing iteration over the contained fields (in index order). Not sure if implementing this on `&Msg` is the right general solution, or how it interacts with `BorrowedMsg`.

2. Add an enum `DecodedField`, which can be decoded into from a field and matched on. See e.g. `field::tests::decode_field`.

### CODE REVIEW

**Comment by bradfier** [comment]
This looks great, I'll review it properly as soon as I can: could you rebase this onto master, it'll get rid of a lot of the noise in the diff introduced by the PRs I've already accepted.

**Comment by pfernie** [comment]
Rebased; I'm not sure if I'm doing the whole GH PR workflow right in cases like this where I have multiple PRs in flight that are inter-related... don't mind rebasing "dependendent" PRs after merging prior if that's the usual thing, but if there is a tidier workflow I'm all ears.

Side note: there is a minor commit in here that was just `rustfmt` changes. Do you have strong opinions on formatting? After these landed, was considering doing a quick `rustfmt` PR, but don't want to create unnecessary churn if there's not an actual policy here...

**Comment by pfernie** [comment]
Oh, I had meant to include in my original PR comment that for the opaque case, `DecodedField` assumes a decode to `u8`, and leaves it to the user to then do other manipulations. Generally, my assumption is for such usage, the user probably isn't using something generic like `DecodedField` as they must have some custom scheme in place. It might even make more sense to simply not support this case for `DecodedField`, it may not have real API benefit.

**Comment by bradfier** [comment]
Yeah I'm not sure I'm a fan of the GH workflow to be perfectly honest, but it is what it is! Thanks for the rebase, I appreciate it can be annoying to base your work off a branch where you've already fixed stuff in another PR.

Has `rustfmt` settled into something stable and sane? It was still in flux last time so I've held off on doing a big reformat. I think my preference is for shorter lines than the default, I like to be able to view two files side-by-side without wrapping, which with line numbers and the gutter realistically means 85-90 chars (I've been using 88 in other projects because why not...) whereas the default is something like 120.

**Comment by bradfier** [comment]
~~And returning a default of `&[u8]` is probably the right thing to do for decoding opaque byte streams. We will assume here that the user must be doing their own ser/de in a format of their choosing, and pretty much everything will accept a slice of bytes.~~

I misread your comment, I don't think it's harmful to return the byte slice as a part of `DecodedField`, it's plausible a user might encode a bunch of byte streams in their own serialization format and want to iterate through them I suppose. I think leave it in.

I seem to recall Rendezvous also has a type for an XML Blob, I didn't implement that since I don't use it anywhere, but if there's a well supported library for an XML Tree maybe we could support decoding into that?

File: src/field.rs
Comment (bradfier) [inline src/field.rs]: While we're discussing style nits and `rustfmt`, if you prefer the lifetime annotations to be 'meaningful' i.e using `'m` or `'msg` instead of `'a` or `'b`, let's reformat all the lifetimes to be meaningful rather than just the new ones.

File: src/field.rs
Comment (bradfier) [inline src/field.rs:78]: I suggest a new `ErrorKind::UnknownFieldType`, it's probably never going to happen so you're right we're better to throw an error than include an `Unknown` variant or such.

Maybe we want to rename `FieldTypeError` to be more specific then?

File: src/message.rs
Comment (bradfier) [inline src/message.rs:46]: > Not sure if implementing this on `&Msg` is the right general solution, or how it interacts with `BorrowedMsg`.

Yep this looks good,, `BorrowedMsg` only exists to represent a nested `Msg` in a field. At the moment basically nothing is implemented for these nested messages, you need to call `.to_owned()` on it and obtain a `Msg` before you can do useful stuff like iterate over the fields. This could probably be improved.

**Review by bradfier — APPROVED** [review]
Aside from the style nits noted, this looks great.

I'm happy to merge now, unless you have any tidying up you would like to do on style or structure?

File: src/message.rs
Comment (bradfier) [reply src/message.rs:46]: Looks like I made a mess of using the review tool!

File: src/field.rs
Comment (pfernie) [reply src/field.rs]: I'm updating this PR to make the style consistent (`'a`); if we'd like to do the meaningful thing, let's keep that in a separate PR. Don't think it has much benefit here; we don't have a lot of confusing lifetimes flying around...

Once I've pushed that, I'm happy w/ this being merged as-is. I will add a couple of issues for other questions/comments in this thread for follow up.

**Pull request merged**
